src/lib/apply-patch/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+//! Applies patches in the [Unified Format](https://www.gnu.org/software/diffutils/manual/html_node/Detailed-Unified.html) to files.
+
+// This will eventually be its own package with more fleshed-out features, but currently just has the bare minimum to implement `dfx new` against it.
+// Missing spots are marked `todo:` with a corresponding `mvp:` section explaining why it doesn't need to be in yet.
+
+// todo: reimplement ::patch. benefits: non-borrowed error, binary patching, consistent newlining, validation of non-overlapping and sortedness.
+// mvp: borrowed errors can be formatted into anyhow, all our patches are text patches to text files in a reasonable format
+use patch::{Line, Patch};
+use thiserror::Error;
+
+/// Applies a single-file patch to `content`.
+///
+/// File paths in the patch file are ignored. Equivalent to `Settings::default().apply_to(patch, content)`.
+pub fn apply_to(patch: &Patch, content: &str) -> Result<String, MismatchError> {
+    Settings::new().apply_to(patch, content)
+}
+
+/// Settings for patch application.
+#[derive(Debug, Clone)]
+pub struct Settings {
+    ignore_line_numbers: bool,
+    ignore_whitespace: bool,
+    // todo: implement multi-patch application to directory. mvp: our patch files are specific
+    _reject_relative_path_segments: bool,
+}
+
+impl Settings {
+    /// Initializes the default settings:
+    ///
+    /// * The line numbers listed in the patch must be the location in the content
+    /// * A whitespace-only difference in the content does not invalidate the patch
+    /// * File paths in multi-file patches are disallowed from containing `..`
+    pub fn new() -> Self {
+        Self {
+            ignore_line_numbers: false,
+            ignore_whitespace: true,
+            _reject_relative_path_segments: true,
+        }
+    }
+    /// Allows line numbers to differ between the patch file and the content.
+    pub fn ignore_line_numbers(self) -> Self {
+        Self {
+            ignore_line_numbers: true,
+            ..self
+        }
+    }
+    /// Requires whitespace to be an exact match between the patch file's context/deleted lines and the content.
+    pub fn exact_whitespace(self) -> Self {
+        Self {
+            ignore_whitespace: false,
+            ..self
+        }
+    }
+    /// Allows `..` in file paths in multi-file patches (not recommended).
+    pub fn allow_relative_path_segments(self) -> Self {
+        Self {
+            _reject_relative_path_segments: false,
+            ..self
+        }
+    }
+    /// Applies a single-file patch to `content`. File paths in the patch file are ignored.
+    // todo: use an iterator instead of returning String. mvp: our files are small.
+    pub fn apply_to(&self, patch: &Patch, content: &str) -> Result<String, MismatchError> {
+        let mut expected_lines = vec![];
+        let mut prev_idx = 0;
```

Comment:
IIUC, `prev_idx` is an offset within `content`.  The data before `prev_idx` has been dealt with (either copied to the output, or deleted), and after is yet to be dealt with.

Suggestions to improve readability:
- rename `content` -> `original`
- rename `new_content` -> `revised`
- One of the following for `prev_idx`:
  - rename `prev_idx` to something like `read_offset`.  Might even be able to get away with `i`.
  - change it to a slice called something like `original_remaining` or even `remaining` that essentially advances through `original`: it's always what `&original[prev_idx..]` would have been.